adapt_common/norms.py

```diff
@@ -0,0 +1,163 @@
+                raise ValueError(val_err) from exc
+            if p < 1:
+                val_err = f"'{norm_type}' norm does not make sense."
+                raise ValueError(val_err)
+            integrand = ufl.inner(v, v)
+        elif norm_type.lower() in ("h1", "hdiv", "hcurl"):
+            integrand = {
+                "h1": lambda w: ufl.inner(w, w) + ufl.inner(ufl.grad(w), ufl.grad(w)),
+                "hdiv": lambda w: ufl.inner(w, w) + ufl.div(w) * ufl.div(w),
+                "hcurl": lambda w: ufl.inner(w, w)
+                + ufl.inner(ufl.curl(w), ufl.curl(w)),
+            }[norm_type.lower()](v)
+        else:
+            val_err = f"Unknown norm type '{norm_type}'."
+            raise ValueError(val_err)
+        return fd.assemble(condition * integrand ** (p / 2) * dX) ** (1 / p)
+
+
+@PETSc.Log.EventDecorator()
+def errornorm(u, uh, norm_type="L2", boundary=False, **kwargs):
+    r"""Overload :func:`fd.norms.errornorm` to allow for :math:`\ell^p` norms.
+
+    Currently supported ``norm_type`` options:
+    * ``'l1'``
+    * ``'l2'``
+    * ``'linf'``
+    * ``'L2'``
+    * ``'Linf'``
+    * ``'H1'``
+    * ``'Hdiv'``
+    * ``'Hcurl'``
+    * or any ``'Lp'`` with :math:`p >= 1`.
+
+    Note that this version is case sensitive, i.e. ``'l2'`` and ``'L2'`` will give
+    different results in general.
+
+    :arg u: the 'true' value
+    :type u: :class:`fd.function.Function` or
+        :class:`fd.cofunction.Cofunction`
+    :arg uh: the approximation of the 'truth'
+    :type uh: :class:`fd.function.Function` or
+        :class:`fd.cofunction.Cofunction`
+    :kwarg norm_type: the type of norm to use
+    :type norm_type: :class:`str`
+    :kwarg boundary: if ``True``, the norm is computed over the domain boundary
+    :type boundary: :class:`bool`
+    :returns: the error norm value
+    :rtype: :class:`float`
+
+    Any other keyword arguments are passed to :func:`fd.norms.errornorm`.
+    """
+    if isinstance(u, fd.Cofunction):
+        u = cofunction2function(u)
+    if isinstance(uh, fd.Cofunction):
+        uh = cofunction2function(uh)
+    if not isinstance(uh, fd.Function):
+        type_err = f"uh should be a Function, is a '{type(uh)}'."
+        raise TypeError(type_err)
+    if norm_type[0] == "l" and not isinstance(u, fd.Function):
+        type_err = f"u should be a Function, is a '{type(u)}'."
+        raise TypeError(type_err)
+
+    if len(u.ufl_shape) != len(uh.ufl_shape):
+        val_err = "Mismatching rank between u and uh."
+        raise ValueError(val_err)
+
+    if isinstance(u, fd.Function):
+        degree_u = u.function_space().ufl_element().degree()
+        degree_uh = uh.function_space().ufl_element().degree()
+        if degree_uh > degree_u:
+            fd.logging.warning(
+                "Degree of exact solution less than approximation degree"
+            )
+
+    # Case 1: point-wise norms
+    if norm_type[0] == "l":
+        v = u
+        v -= uh
```

Comment:
If I'm not mistaken `-=` for Functions is overloaded and basically does `v.assign(v -uh)` which again means that the values of the input u are overwritten as v is the same as u at this point. So then if you call this errornorm twice with the same u and uh, you get the wrong thing on the 2nd call.

Reply:
Another good point. Reworked `errornorm` in 9664a24.